src/compiler/preprocess/index.ts

```diff
@@ -1,218 +1,168 @@
-	const replacements: Array<Promise<Replacement>> = [];
-	source.replace(re, (...args) => {
-		replacements.push(
-			func(...args).then(
-				res =>
-					({
-						offset: args[args.length - 2],
-						length: args[0].length,
-						replacement: res
-					}) as Replacement
-			)
-		);
-		return '';
-	});
-	const out = new StringWithSourcemap();
-	let last_end = 0;
-	for (const { offset, length, replacement } of await Promise.all(
-		replacements
-	)) {
-		// content = unchanged source characters before the replaced segment
-		const content = StringWithSourcemap.from_source(
-			file_basename, source.slice(last_end, offset), get_location(last_end));
-		out.concat(content).concat(replacement);
-		last_end = offset + length;
+	get_location: ReturnType<typeof getLocator>;
+
+	constructor(public source: string, public filename: string) {
+		this.update_source({ string: source });
+
+		// preprocess source must be relative to itself or equal null
+		this.file_basename = filename == null ? null : get_file_basename(filename);
 	}
-	// final_content = unchanged source characters after last replaced segment
-	const final_content = StringWithSourcemap.from_source(
-		file_basename, source.slice(last_end), get_location(last_end));
-	return out.concat(final_content);
-}
 
-/**
- * Import decoded sourcemap from mozilla/source-map/SourceMapGenerator
- * Forked from source-map/lib/source-map-generator.js
- * from methods _serializeMappings and toJSON.
- * We cannot use source-map.d.ts types, because we access hidden properties.
- */
-function decoded_sourcemap_from_generator(generator: any) {
-	let previous_generated_line = 1;
-	const converted_mappings = [[]];
-	let result_line;
-	let result_segment;
-	let mapping;
-
-	const source_idx = generator._sources.toArray()
-		.reduce((acc, val, idx) => (acc[val] = idx, acc), {});
-
-	const name_idx = generator._names.toArray()
-		.reduce((acc, val, idx) => (acc[val] = idx, acc), {});
-
-	const mappings = generator._mappings.toArray();
-	result_line = converted_mappings[0];
-
-	for (let i = 0, len = mappings.length; i < len; i++) {
-		mapping = mappings[i];
-
-		if (mapping.generatedLine > previous_generated_line) {
-			while (mapping.generatedLine > previous_generated_line) {
-				converted_mappings.push([]);
-				previous_generated_line++;
-			}
-			result_line = converted_mappings[mapping.generatedLine - 1]; // line is one-based
-		} else if (i > 0) {
-			const previous_mapping = mappings[i - 1];
-			if (
-				// sorted by selectivity
-				mapping.generatedColumn === previous_mapping.generatedColumn &&
-				mapping.originalColumn === previous_mapping.originalColumn &&
-				mapping.name === previous_mapping.name &&
-				mapping.generatedLine === previous_mapping.generatedLine &&
-				mapping.originalLine === previous_mapping.originalLine &&
-				mapping.source === previous_mapping.source
-		) {
-				continue;
-			}
+	update_source({ string: source, map, dependencies }: SourceUpdate) {
+		this.source = source;
```

Comment:
hmm... didnt see you define `this.source` in `PreprocessResult`

Reply:
it's done by the `public` keyword in the constructor: `constructor(public source: string, public filename: string)`